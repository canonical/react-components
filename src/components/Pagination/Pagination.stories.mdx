import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";

import Pagination from "./Pagination";

<Meta title="Pagination" component={Pagination} />

export const Template = (args) => <Pagination {...args} />;

### Pagination

This is a [React](https://reactjs.org/) component for the Vanilla [Pagination](https://docs.vanillaframework.io/patterns/pagination/).

The pagination component should be used to navigate between pages of content. Depending on the length provided, the pagination component will automatically scale.

### Props

<ArgsTable of={Pagination} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      itemsPerPage: 20,
      totalItems: 100,
      paginate: () => {},
      currentPage: 3,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

### Truncated

<Canvas>
  <Story name="Truncated">
    <Pagination
      itemsPerPage={10}
      totalItems={1000}
      paginate={() => {}}
      currentPage={34}
    />
  </Story>
</Canvas>

### Disabled controls

<Canvas>
  <Story name="Disabled controls">
    <Pagination
      itemsPerPage={20}
      totalItems={100}
      paginate={() => {}}
      currentPage={1}
    />
  </Story>
</Canvas>

### Centered

<Canvas>
  <Story name="Centered">
    <Pagination
      itemsPerPage={10}
      totalItems={50}
      paginate={() => {}}
      currentPage={1}
      centered
    />
  </Story>
</Canvas>
