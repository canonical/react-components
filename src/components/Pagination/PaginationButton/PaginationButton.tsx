import classNames from "classnames";
import React from "react";
import type { MouseEventHandler } from "react";

export type PaginationDirection = "forward" | "back";
export type Props = {
  /**
   * The direction of the pagination.
   */
  direction: PaginationDirection;
  /**
   * Whether the pagination button should be disabled.
   */
  disabled?: boolean;
  /**
   * Function to handle clicking the pagination button.
   */
  onClick: MouseEventHandler<HTMLButtonElement>;
};

const PaginationButton = ({
  direction,
  onClick,
  disabled = false,
}: Props): JSX.Element => {
  const label = direction === "back" ? "Previous page" : "Next page";
  return (
    <li className="p-pagination__item">
      <button
        className={classNames({
          "p-pagination__link--previous": direction === "back",
          "p-pagination__link--next": direction === "forward",
        })}
        disabled={disabled}
        onClick={onClick}
      >
        <i className="p-icon--chevron-down">{label}</i>
      </button>
    </li>
  );
};

export default PaginationButton;
