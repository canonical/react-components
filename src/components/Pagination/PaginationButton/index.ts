export { default } from "./PaginationButton";
export type { Props as PaginationButtonProps } from "./PaginationButton";
