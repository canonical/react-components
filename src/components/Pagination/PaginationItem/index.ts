export { default } from "./PaginationItem";
export type { Props as PaginationItemProps } from "./PaginationItem";
