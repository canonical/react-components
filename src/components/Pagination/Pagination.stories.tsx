import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Pagination from "./Pagination";

const meta: Meta<typeof Pagination> = {
  component: Pagination,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Pagination>;

export const Default: Story = {
  name: "Default",

  args: {
    itemsPerPage: 20,
    totalItems: 100,
    paginate: () => {},
    currentPage: 3,
  },
};

export const TruncatedNumberedPagination: Story = {
  render: () => (
    <Pagination
      itemsPerPage={10}
      totalItems={1000}
      paginate={() => {}}
      currentPage={34}
    />
  ),
  name: "Truncated (Numbered Pagination)",
};

export const DisabledControlsNumberedPagination: Story = {
  render: () => (
    <Pagination
      itemsPerPage={20}
      totalItems={100}
      paginate={() => {}}
      currentPage={1}
    />
  ),
  name: "Disabled controls (Numbered Pagination)",
};

export const CenteredNumberedPagination: Story = {
  render: () => (
    <Pagination
      itemsPerPage={10}
      totalItems={50}
      paginate={() => {}}
      currentPage={1}
      centered
    />
  ),

  name: "Centered (Numbered Pagination)",
};

export const DisabledPaginationItemsNumberedPagination: Story = {
  render: () => (
    <Pagination
      itemsPerPage={10}
      totalItems={50}
      paginate={() => {}}
      currentPage={1}
      hideNumbers
    />
  ),

  name: "Disabled pagination items (Numbered Pagination)",
};

export const DefaultButtonsOnlyPagination: Story = {
  render: () => <Pagination onForward={() => {}} onBack={() => {}} />,
  name: "Default (Buttons-only Pagination)",
};

export const DefaultButtonLabelsButtonsOnlyPagination: Story = {
  render: () => (
    <Pagination onForward={() => {}} onBack={() => {}} showLabels />
  ),
  name: "Default button labels (Buttons-only Pagination)",
};

export const CustomButtonLabelsButtonsOnlyPagination: Story = {
  render: () => (
    <Pagination
      onForward={() => {}}
      onBack={() => {}}
      showLabels
      forwardLabel="Custom forward label"
      backLabel="Custom back label"
    />
  ),

  name: "Custom button labels (Buttons-only Pagination)",
};

export const CenteredButtonsOnlyPagination: Story = {
  render: () => <Pagination onForward={() => {}} onBack={() => {}} centered />,
  name: "Centered (Buttons-only Pagination)",
};
