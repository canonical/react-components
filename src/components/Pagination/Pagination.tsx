import classNames from "classnames";
import React from "react";
import type { HTMLProps } from "react";

import PaginationButton from "./PaginationButton";
import PaginationItem from "./PaginationItem";

import type { PropsWithSpread } from "types";
import type { PaginationDirection } from "./PaginationButton/PaginationButton";

const scrollTop = () => window.scrollTo(0, 0);

const generatePaginationItems = (
  pageNumbers: number[],
  currentPage: number,
  truncateThreshold: number,
  changePage: (page: number) => void
) => {
  const lastPage = pageNumbers.length;
  const truncated = lastPage > truncateThreshold;

  let visiblePages: number[];
  if (truncated) {
    // the default range for pages outside the start and end threshold
    let start = currentPage - 2;
    let end = currentPage + 1;
    // on page 1, also show pages 2, 3 and 4
    if (currentPage === 1) {
      start = 1;
      end = Math.min(lastPage - 1, currentPage + 3);
    }
    // on page 2, show page 1, and also pages 3, and 4
    if (currentPage === 2) {
      start = 1;
      end = Math.min(lastPage - 1, currentPage + 3);
    }
    // on the last page and page before last, also show the 3 previous pages
    if (currentPage === lastPage || currentPage === lastPage - 1) {
      start = Math.max(1, lastPage - 4);
      end = lastPage - 1;
    }
    visiblePages = pageNumbers.slice(start, end);
  } else {
    visiblePages = pageNumbers;
  }

  const items = [];
  if (truncated) {
    // render first in sequence
    items.push(
      <PaginationItem
        key={1}
        number={1}
        isActive={currentPage === 1}
        onClick={() => changePage(1)}
      />
    );
    if (!visiblePages.includes(2)) {
      items.push(<PaginationItemSeparator key="sep1" />);
    }
  }

  items.push(
    visiblePages.map((number) => (
      <PaginationItem
        key={number}
        number={number}
        isActive={number === currentPage}
        onClick={() => changePage(number)}
      />
    ))
  );

  if (truncated) {
    // render last in sequence
    if (!visiblePages.includes(lastPage - 1)) {
      items.push(<PaginationItemSeparator key="sep2" />);
    }
    items.push(
      <PaginationItem
        key={lastPage}
        number={lastPage}
        isActive={currentPage === lastPage}
        onClick={() => changePage(lastPage)}
      />
    );
  }
  return items;
};

const PaginationItemSeparator = (): JSX.Element => (
  <li className="p-pagination__item p-pagination__item--truncation">
    &hellip;
  </li>
);

// Props that are used in both types of paginations.
type BaseProps = {
  /**
   * Whether to scroll to the top of the list on page change.
   */
  scrollToTop?: boolean;
  /**
   * The number of pages at which to truncate the pagination items.
   */
  truncateThreshold?: number;
  /**
   * Whether the pagination is ceneterd on the page.
   */
  centered?: boolean;
  /**
   * Whether to show the labels for forward and back buttons.
   */
  showLabels?: boolean;
  /**
   * Whether forward button is disabled.
   */
  forwardDisabled?: boolean;
  /**
   * Whether back button is disabled.
   */
  backDisabled?: boolean;
  /**
   * Custom label for forward button.
   */
  forwardLabel?: string;
  /**
   * Custom label for back button.
   */
  backLabel?: string;
};

// Used when number of items per page, number of total items,
// current page and paginate function is defined. Pagination
// is handled by paginate function and optional onForward and
// onBack functions.
type NumberedPagination = BaseProps & {
  /**
   * The current page being viewed.<br>
   * **Required for Numbered Pagination.**
   */
  currentPage: number;
  /**
   * The number of items to show per page.<br>
   * **Required for numbered pagination.**
   */
  itemsPerPage: number;
  /**
   * Function to handle paginating the items.<br>
   * **Required for numbered pagination.**
   */
  paginate: (page: number) => void;
  /**
   * The total number of items.<br>
   * **Required for numbered pagination.**
   */
  totalItems: number;
  /**
   * Whether to hide the pagination items.
   */
  hideNumbers?: boolean;
  /**
   * Function to handle page transition to a higher-numbered page for
   * numbered pagination and to next page for buttons-only pagination.<br>
   * _Called with page parameter for numbered pagination and_
   * _with no parameter for buttons-only pagination._<br>
   * **Required for buttons-only pagination.**
   */
  onForward?: (page: number) => void;
  /**
   * Function to handle page transition to a lower-numbered page for
   * numbered pagination and to next page for buttons-only pagination.<br>
   * _Called with page parameter for numbered pagination and_
   * _with no parameter for buttons-only pagination._<br>
   * **Required for buttons-only pagination.**
   */
  onBack?: (page: number) => void;
};

// Used when number of items per page, number of total items,
// current page and paginate function are undefined.
// Pagination is handled by onForward and onBack function.
type ButtonsOnlyPagination = BaseProps & {
  itemsPerPage?: never;
  totalItems?: never;
  currentPage?: never;
  paginate?: never;
  hideNumbers?: never;
  onForward: () => void;
  onBack: () => void;
};

export type Props = PropsWithSpread<
  NumberedPagination | ButtonsOnlyPagination,
  HTMLProps<HTMLElement>
>;

/**
This is a [React](https://reactjs.org/) component for the Vanilla [Pagination](https://docs.vanillaframework.io/patterns/pagination/).

The pagination component should be used to navigate between pages of content. There are two types of pagination being provided:

* **Numbered Pagination:** used when the total number of items is known.<br />
  Depending on the length provided, the pagination component will automatically scale.<br />
  **Required props:** itemsPerPage, totalItems, paginate, currentPage.

* **Buttons-only Pagination:** used when the total number of items is unknown.<br />
  **Required props:** onForward, onBack.
 */
const Pagination = ({
  itemsPerPage,
  totalItems,
  paginate,
  currentPage,
  scrollToTop,
  truncateThreshold = 10,
  centered,
  showLabels,
  hideNumbers,
  onForward,
  onBack,
  forwardDisabled,
  backDisabled,
  forwardLabel,
  backLabel,
  ...navProps
}: Props): JSX.Element => {
  const isNumberedPagination =
    !!itemsPerPage && !!totalItems && !!currentPage && !!paginate;

  const pageNumbers = [];

  if (isNumberedPagination) {
    // return early if no pagination is required
    if (totalItems <= itemsPerPage) {
      return null;
    }

    for (let i = 1; i <= Math.ceil(totalItems / itemsPerPage); i++) {
      pageNumbers.push(i);
    }
  }

  const changeNumberedPage = (page: number) => {
    paginate(page);
    if (page > currentPage) {
      onForward?.(page);
    }
    if (page < currentPage) {
      onBack?.(page);
    }
    scrollToTop && scrollTop();
  };

  const changeButtonsOnlyPage = (direction: PaginationDirection) => {
    if (direction === "forward") {
      onForward?.(undefined);
    }
    if (direction === "back") {
      onBack?.(undefined);
    }
    scrollToTop && scrollTop();
  };

  return (
    <nav className="p-pagination" aria-label="Pagination" {...navProps}>
      <ol
        className={classNames("p-pagination__items", {
          "u-align--center": centered,
        })}
      >
        <PaginationButton
          key="back"
          direction="back"
          disabled={backDisabled || currentPage === 1}
          onClick={() =>
            isNumberedPagination
              ? changeNumberedPage(currentPage - 1)
              : changeButtonsOnlyPage("back")
          }
          showLabel={showLabels}
          label={backLabel}
        />
        {isNumberedPagination && !hideNumbers
          ? generatePaginationItems(
              pageNumbers,
              currentPage,
              truncateThreshold,
              changeNumberedPage
            )
          : null}
        <PaginationButton
          key="forward"
          direction="forward"
          disabled={forwardDisabled || currentPage === pageNumbers.length}
          onClick={() =>
            isNumberedPagination
              ? changeNumberedPage(currentPage + 1)
              : changeButtonsOnlyPage("forward")
          }
          showLabel={showLabels}
          label={forwardLabel}
        />
      </ol>
    </nav>
  );
};

export default Pagination;
