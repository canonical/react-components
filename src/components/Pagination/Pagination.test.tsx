import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Pagination from "./Pagination";
import { Label as PaginationButtonLabel } from "./PaginationButton/PaginationButton";

describe("<Pagination />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    render(
      <Pagination
        itemsPerPage={10}
        totalItems={50}
        paginate={jest.fn()}
        currentPage={1}
      />
    );

    expect(screen.getByRole("navigation")).toMatchSnapshot();
  });

  // unit tests
  it("renders no pagination with only a single page", () => {
    render(
      <Pagination
        itemsPerPage={10}
        totalItems={5}
        paginate={jest.fn()}
        currentPage={1}
      />
    );

    expect(screen.queryByRole("navigation")).not.toBeInTheDocument();
  });

  it("renders a simple paginator with back and forward arrows if only five pages or less", () => {
    render(
      <Pagination
        itemsPerPage={10}
        totalItems={50}
        paginate={jest.fn()}
        currentPage={1}
      />
    );

    expect(
      screen.queryByRole("listitem", { name: "…" })
    ).not.toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: PaginationButtonLabel.Next })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: PaginationButtonLabel.Previous })
    ).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "1" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "2" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "3" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "4" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "5" })).toBeInTheDocument();
  });

  it("renders a complex paginator with truncation if more than five pages", () => {
    render(
      <Pagination
        itemsPerPage={10}
        totalItems={1000}
        paginate={jest.fn()}
        currentPage={5}
      />
    );

    expect(screen.getAllByText("…")).toHaveLength(2);
    expect(
      screen.getByRole("button", { name: PaginationButtonLabel.Next })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: PaginationButtonLabel.Previous })
    ).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "1" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "4" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "5" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "6" })).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "100" })).toBeInTheDocument();
  });

  it("does not render a truncation separator if currentPage is contiguous at start", () => {
    render(
      <Pagination
        itemsPerPage={10}
        totalItems={1000}
        paginate={jest.fn()}
        currentPage={2}
      />
    );

    // There should only be one ellipsis.
    expect(screen.getAllByText("…")).toHaveLength(1);
  });

  it("does not render a truncation separator if currentPage is contiguous at end", () => {
    render(
      <Pagination
        itemsPerPage={10}
        totalItems={1000}
        paginate={jest.fn()}
        currentPage={98}
      />
    );

    // There should only be one ellipsis.
    expect(screen.getAllByText("…")).toHaveLength(1);
  });

  it("does not trigger form submission on pagination button click by default", async () => {
    const handleOnSubmit = jest.fn();
    render(
      <form onSubmit={handleOnSubmit}>
        <Pagination
          itemsPerPage={10}
          totalItems={1000}
          paginate={jest.fn()}
          currentPage={98}
        />
      </form>
    );

    await userEvent.click(screen.getByRole("button", { name: "Next page" }));
    await userEvent.click(screen.getByRole("button", { name: "99" }));
    expect(handleOnSubmit).not.toHaveBeenCalled();
  });

  it("can be centered", () => {
    render(
      <Pagination
        itemsPerPage={10}
        totalItems={1000}
        paginate={jest.fn()}
        currentPage={98}
        centered
      />
    );
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(".p-pagination")).toHaveClass(
      "u-align--center"
    );
  });
});
