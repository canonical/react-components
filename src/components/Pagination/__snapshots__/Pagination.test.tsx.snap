// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders and matches the snapshot 1`] = `
<nav>
  <ul
    class="p-pagination"
  >
    <li
      class="p-pagination__item"
    >
      <button
        class="p-pagination__link--previous"
        disabled=""
        type="button"
      >
        <i
          class="p-icon--chevron-down"
        >
          Previous page
        </i>
      </button>
    </li>
    <li
      class="p-pagination__item"
    >
      <button
        aria-current="page"
        class="p-pagination__link is-active"
        type="button"
      >
        1
      </button>
    </li>
    <li
      class="p-pagination__item"
    >
      <button
        class="p-pagination__link"
        type="button"
      >
        2
      </button>
    </li>
    <li
      class="p-pagination__item"
    >
      <button
        class="p-pagination__link"
        type="button"
      >
        3
      </button>
    </li>
    <li
      class="p-pagination__item"
    >
      <button
        class="p-pagination__link"
        type="button"
      >
        4
      </button>
    </li>
    <li
      class="p-pagination__item"
    >
      <button
        class="p-pagination__link"
        type="button"
      >
        5
      </button>
    </li>
    <li
      class="p-pagination__item"
    >
      <button
        class="p-pagination__link--next"
        type="button"
      >
        <i
          class="p-icon--chevron-down"
        >
          Next page
        </i>
      </button>
    </li>
  </ul>
</nav>
`;
